### ISSUE TITLE
Autocomplete "__" after dot if all methods starts with __

### ISSUE THREAD

**Issue by ontaptom** [davidhalter/jedi-vim#868]
### Issue

Hi, I'm having issue with jedi-vim  that tries to autocomplete the method name for me, if it "discovers" that all methods starts with __.  Shortest example I could think of, when I type:

```
class MyClass():
    def __init__(self,myname):
        self
```

Right after putting "." I will imidietaly have:

```
class MyClass():
    def __init__(self,myname):
        self.__
```

Of course I never actually typed  "__"! The annoying thing is that it actually "autocompletes" even if I copy this code:   

`self.myname = myname`

and paste that in this place of code, it got changed to:

`self.__myname = myname`

I really don't know how to stop it, and it seems I've found same issue reported previously (#429), however the solution was not clear.  I have installed jedi-vim via pathogen.

My completeopt:
` completeopt=menuone,longest,preview `
If I apply in my .vimrc:
```
let g:jedi#auto_vim_configuration = 0
let g:jedi#popup_select_first = 0
set completeopt=menuone
```
Behaviour is still the same. Removing basically everything from .vimrc (besides  execute pathogen#infect()) also did not changed the behaviour.

### Steps to reproduce

Open new .py file and type
```
class MyClass():
    def __init__(self,myname):
        self
```
And press add "." after self

Same behaviour reproduced with minimal.vimrc
```
set nocompatible

let script_dir = fnamemodify(expand('<sfile>'), ':h')
let &runtimepath .= ','.script_dir.','.script_dir.'/after'
syntax on
filetype plugin indent on
```
##### jedi-vim version

 - jedi-vim git version: Unknown option: -C
usage: git [--version] [--help] [-c name=value]
           [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p|--paginate|--no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           <command> [<args>]
 - jedi git submodule status: Unknown option: -C
usage: git [--version] [--help] [-c name=value]
           [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p|--paginate|--no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           <command> [<args>]
 - parso git submodule status: Unknown option: -C
usage: git [--version] [--help] [-c name=value]
           [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p|--paginate|--no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           <command> [<args>]

##### Global Python

Using Python version 2 to access Jedi.
 - global sys.version: `2.7.5 (default, Jul 13 2018, 13:06:57), [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]`
 - global site module: `/usr/lib64/python2.7/site.pyc`

##### Jedi

 - path: `/home/user/.vim/bundle/jedi-vim/pythonx/jedi/jedi/__init__.pyc`
 - version: 0.12.1

##### Jedi environment: <Environment: 3.6.5 in /usr/local>

 - executable: /usr/local/bin/python3.6
 - sys_path:
    - `/usr/local/lib/python36.zip`
    - `/usr/local/lib/python3.6`
    - `/usr/local/lib/python3.6/lib-dynload`
    - `/home/user/.local/lib/python3.6/site-packages`
    - `/usr/local/lib/python3.6/site-packages`

##### Known environments

 - <Environment: 3.6.5 in /usr/local> (/usr/local/bin/python3.6)

 - <Environment: 2.7.5 in /usr> (/usr/bin/python2.7)

##### Settings

```
  omnifunc=jedi#completions
        Last set from ~/.vim/bundle/jedi-vim/autoload/jedi.vim
  completeopt=menuone,longest,preview
        Last set from ~/.vim/bundle/jedi-vim/plugin/jedi.vim
```

#### :version
```

VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Apr 10 2018 23:54:40)
Included patches: 1-160, 399, 402-403, 1099
Modified by <bugzilla@redhat.com>
Compiled by <bugzilla@redhat.com>
Huge version without GUI.  Features included (+) or not (-):
+acl             +conceal         +farsi           +libcall         +mouse_netterm   +profile         +syntax          +visualextra
+arabic          +cryptv          +file_in_path    +linebreak       +mouse_sgr       +python/dyn      +tag_binary      +viminfo
+autocmd         +cscope          +find_in_path    +lispindent      -mouse_sysmouse  -python3         +tag_old_static  +vreplace
-balloon_eval    +cursorbind      +float           +listcmds        +mouse_urxvt     +quickfix        -tag_any_white   +wildignore
-browse          +cursorshape     +folding         +localmap        +mouse_xterm     +reltime         -tcl             +wildmenu
++builtin_terms  +dialog_con      -footer          -lua             +multi_byte      +rightleft       +terminfo        +windows
+byte_offset     +diff            +fork()          +menu            +multi_lang      +ruby/dyn        +termresponse    +writebackup
+cindent         +digraphs        +gettext         +mksession       -mzscheme        +scrollbind      +textobjects     -X11
-clientserver    -dnd             -hangul_input    +modify_fname    +netbeans_intg   +signs           +title           -xfontset
-clipboard       -ebcdic          +iconv           +mouse           +path_extra      +smartindent     -toolbar         -xim
+cmdline_compl   +emacs_tags      +insert_expand   -mouseshape      +perl            -sniff           +user_commands   -xsmp
+cmdline_hist    +eval            +jumplist        +mouse_dec       +persistent_undo +startuptime     +vertsplit       -xterm_clipboard
+cmdline_info    +ex_extra        +keymap          +mouse_gpm       +postscript      +statusline      +virtualedit     -xterm_save
+comments        +extra_search    +langmap         -mouse_jsbterm   +printer         -sun_workshop    +visual          -xpm
   system vimrc file: "/etc/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  fall-back for $VIM: "/etc"
 f-b for $VIMRUNTIME: "/usr/share/vim/vim74"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H     -O2 -g -pipe -Wall -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
Linking: gcc   -L. -Wl,-z,relro -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,--enable-new-dtags -Wl,-rpath,/usr/lib64/perl5/CORE  -Wl,-z,relro  -L/usr/local/lib -Wl,--as-needed -o vim        -lm -lnsl  -lselinux  -lncurses -lacl -lattr -lgpm -ldl   -Wl,--enable-new-dtags -Wl,-rpath,/usr/lib64/perl5/CORE  -fstack-protector  -L/usr/lib64/perl5/CORE -lperl -lresolv -lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
```

#### :messages
```
Messages maintainer: Bram Moolenaar <Bram@vim.org>
"code/test.py" 4L, 60C
E475: Invalid argument: @+>
E354: Invalid register name: '+'

1 change; before #3  2 seconds ago
E475: Invalid argument:
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Apr 10 2018 23:54:40)
  1: /etc/vimrc
  2: /usr/share/vim/vim74/syntax/syntax.vim
  3: /usr/share/vim/vim74/syntax/synload.vim
  4: /usr/share/vim/vim74/syntax/syncolor.vim
  5: /usr/share/vim/vim74/filetype.vim
  6: /usr/share/vim/vimfiles/ftdetect/stp.vim
  7: /usr/share/vim/vim74/ftplugin.vim
  8: ~/.vimrc
  9: ~/.vim/autoload/pathogen.vim
 10: /usr/share/vim/vim74/ftoff.vim
 11: ~/.vim/bundle/vim-colorschemes/colors/gruvbox.vim
 12: /usr/share/vim/vim74/indent.vim
 13: ~/.vim/bundle/jedi-vim/plugin/jedi.vim
 14: ~/.vim/bundle/nerdtree/plugin/NERD_tree.vim
 15: ~/.vim/bundle/nerdtree/autoload/nerdtree.vim
 16: ~/.vim/bundle/nerdtree/lib/nerdtree/path.vim
 17: ~/.vim/bundle/nerdtree/lib/nerdtree/menu_controller.vim
 18: ~/.vim/bundle/nerdtree/lib/nerdtree/menu_item.vim
 19: ~/.vim/bundle/nerdtree/lib/nerdtree/key_map.vim
 20: ~/.vim/bundle/nerdtree/lib/nerdtree/bookmark.vim
 21: ~/.vim/bundle/nerdtree/lib/nerdtree/tree_file_node.vim
 22: ~/.vim/bundle/nerdtree/lib/nerdtree/tree_dir_node.vim
 23: ~/.vim/bundle/nerdtree/lib/nerdtree/opener.vim
 24: ~/.vim/bundle/nerdtree/lib/nerdtree/creator.vim
 25: ~/.vim/bundle/nerdtree/lib/nerdtree/flag_set.vim
 26: ~/.vim/bundle/nerdtree/lib/nerdtree/nerdtree.vim
 27: ~/.vim/bundle/nerdtree/lib/nerdtree/ui.vim
 28: ~/.vim/bundle/nerdtree/lib/nerdtree/event.vim
 29: ~/.vim/bundle/nerdtree/lib/nerdtree/notifier.vim
 30: ~/.vim/bundle/nerdtree/autoload/nerdtree/ui_glue.vim
 31: ~/.vim/bundle/nerdtree/nerdtree_plugin/exec_menuitem.vim
 32: ~/.vim/bundle/nerdtree/nerdtree_plugin/fs_menu.vim
 33: ~/.vim/bundle/setcolors/plugin/setcolors.vim
 34: /usr/share/vim/vimfiles/plugin/cctree.vim
 35: /usr/share/vim/vim74/plugin/getscriptPlugin.vim
 36: /usr/share/vim/vim74/plugin/gzip.vim
 37: /usr/share/vim/vim74/plugin/matchparen.vim
 38: /usr/share/vim/vim74/plugin/netrwPlugin.vim
 39: /usr/share/vim/vim74/plugin/rrhelper.vim
 40: /usr/share/vim/vim74/plugin/spellfile.vim
 41: /usr/share/vim/vim74/plugin/tarPlugin.vim
 42: /usr/share/vim/vim74/plugin/tohtml.vim
 43: /usr/share/vim/vim74/plugin/vimballPlugin.vim
 44: /usr/share/vim/vim74/plugin/zipPlugin.vim
 45: /usr/share/vim/vim74/syntax/python.vim
 46: ~/.vim/bundle/jedi-vim/after/syntax/python.vim
 47: ~/.vim/bundle/jedi-vim/autoload/jedi.vim
 48: ~/.vim/bundle/jedi-vim/ftplugin/python/jedi.vim
 49: /usr/share/vim/vim74/ftplugin/python.vim
 50: ~/.vim/bundle/jedi-vim/after/ftplugin/python/jedi.vim
 51: /usr/share/vim/vim74/indent/python.vim
```

<details><summary>:scriptnames</summary>

```
  1: /etc/vimrc
  2: /usr/share/vim/vim74/syntax/syntax.vim
  3: /usr/share/vim/vim74/syntax/synload.vim
  4: /usr/share/vim/vim74/syntax/syncolor.vim
  5: /usr/share/vim/vim74/filetype.vim
  6: /usr/share/vim/vimfiles/ftdetect/stp.vim
  7: /usr/share/vim/vim74/ftplugin.vim
  8: ~/.vimrc
  9: ~/.vim/autoload/pathogen.vim
 10: /usr/share/vim/vim74/ftoff.vim
 11: ~/.vim/bundle/vim-colorschemes/colors/gruvbox.vim
 12: /usr/share/vim/vim74/indent.vim
 13: ~/.vim/bundle/jedi-vim/plugin/jedi.vim
 14: ~/.vim/bundle/nerdtree/plugin/NERD_tree.vim
 15: ~/.vim/bundle/nerdtree/autoload/nerdtree.vim
 16: ~/.vim/bundle/nerdtree/lib/nerdtree/path.vim
 17: ~/.vim/bundle/nerdtree/lib/nerdtree/menu_controller.vim
 18: ~/.vim/bundle/nerdtree/lib/nerdtree/menu_item.vim
 19: ~/.vim/bundle/nerdtree/lib/nerdtree/key_map.vim
 20: ~/.vim/bundle/nerdtree/lib/nerdtree/bookmark.vim
 21: ~/.vim/bundle/nerdtree/lib/nerdtree/tree_file_node.vim
 22: ~/.vim/bundle/nerdtree/lib/nerdtree/tree_dir_node.vim
 23: ~/.vim/bundle/nerdtree/lib/nerdtree/opener.vim
 24: ~/.vim/bundle/nerdtree/lib/nerdtree/creator.vim
 25: ~/.vim/bundle/nerdtree/lib/nerdtree/flag_set.vim
 26: ~/.vim/bundle/nerdtree/lib/nerdtree/nerdtree.vim
 27: ~/.vim/bundle/nerdtree/lib/nerdtree/ui.vim
 28: ~/.vim/bundle/nerdtree/lib/nerdtree/event.vim
 29: ~/.vim/bundle/nerdtree/lib/nerdtree/notifier.vim
 30: ~/.vim/bundle/nerdtree/autoload/nerdtree/ui_glue.vim
 31: ~/.vim/bundle/nerdtree/nerdtree_plugin/exec_menuitem.vim
 32: ~/.vim/bundle/nerdtree/nerdtree_plugin/fs_menu.vim
 33: ~/.vim/bundle/setcolors/plugin/setcolors.vim
 34: /usr/share/vim/vimfiles/plugin/cctree.vim
 35: /usr/share/vim/vim74/plugin/getscriptPlugin.vim
 36: /usr/share/vim/vim74/plugin/gzip.vim
 37: /usr/share/vim/vim74/plugin/matchparen.vim
 38: /usr/share/vim/vim74/plugin/netrwPlugin.vim
 39: /usr/share/vim/vim74/plugin/rrhelper.vim
 40: /usr/share/vim/vim74/plugin/spellfile.vim
 41: /usr/share/vim/vim74/plugin/tarPlugin.vim
 42: /usr/share/vim/vim74/plugin/tohtml.vim
 43: /usr/share/vim/vim74/plugin/vimballPlugin.vim
 44: /usr/share/vim/vim74/plugin/zipPlugin.vim
 45: /usr/share/vim/vim74/syntax/python.vim
 46: ~/.vim/bundle/jedi-vim/after/syntax/python.vim
 47: ~/.vim/bundle/jedi-vim/autoload/jedi.vim
 48: ~/.vim/bundle/jedi-vim/ftplugin/python/jedi.vim
 49: /usr/share/vim/vim74/ftplugin/python.vim
 50: ~/.vim/bundle/jedi-vim/after/ftplugin/python/jedi.vim
 51: /usr/share/vim/vim74/indent/python.vim
```
</details>

**Comment by davidhalter** [comment]
Please fill out the form to have enough information to debug your issue. You have intentionally ignored it, by deleting the form when creating a new issue. I'm just pointing it out, because the form is there for a reason. It saves us a lot of time and questions if we have it.

We are happy to reopen this issue if you add the necessary information.

**Comment by ontaptom** [comment]
Hi David, thanks for taking the interest, and apologies for lack of details in the first place.
I couldn't execute `:redir @+> | silent verb JediDebugInfo | redir END` since that reported E475: Invalid argument: @+>,    I have modified it to redir @a, I hope it's not relevant.

**Comment by davidhalter** [comment]
Thanks for adding the necessary information!

It's strange, I cannot reproduce your issue with a very similar setup. Since `git -C` does not work, could you check these repositories quickly and tell us what the versions/git commits in use are?

**Comment by ontaptom** [comment]
Sure!
$ git --version
git version 1.8.3.1

$ git submodule status
 a79a1fbef57564d031c59f47def20a6d9adf3439 pythonx/jedi (v0.12.1)
 e6bc924fbabbfa3090044e475c91a7655b9c6e9b pythonx/parso (v0.3.1)

**Comment by ontaptom** [comment]
Yes, also jedi-vim master, latest. I tried to install supertab, but that seems to have no impact. However on another host, with Vim 8.1 and exactly the same version of jedi-vim the problem does not occur, so it seems that might be related strictly with vim itself. A bit off.

**Comment by davidhalter** [comment]
Ok. I just realized what the issue is: You need `completeopt+=noinsert` capability to make it work. We filter for that with `if has('patch-7.4-775')`. Since your older VIM doesn't include that patch, we cannot give you the ability for that.

I will close with the recommendation to download a newer VIM and thank @blueyed again for fixing this in newer VIM versions :).

**Comment by blueyed** [comment]
@ontaptom 
I guess `set completeopt-=longest` makes it better for you?

I've not really investigated, but maybe we should remove this code?

https://github.com/davidhalter/jedi-vim/blob/ed8cfbd57e56586f7d552cd58a44566826c1b27d/autoload/jedi.vim#L721-L727

**Comment by blueyed** [comment]
@ontaptom 
jedi-vim changes `completeopt` temporarily, as long as you are not using `noinsert`/`noselect`: https://github.com/davidhalter/jedi-vim/blob/ed8cfbd57e56586f7d552cd58a44566826c1b27d/autoload/jedi.vim#L707-L728

Please try if removing the `set completeopt+=longest` (line 726) makes it behave saner.

**Comment by ontaptom** [comment]
Thanks @Blueyed, I tried it and it definitely changed the behavior, however for a bit worse :).
After removing `set completeopt+=longest  from code, in vim7.4 the first option was automatically completed, so in that particular case 

`self.` got autocompleted to `self.__class__`. 

I tried to play with 
`let g:jedi#popup_select_first = 0` and tried to modify `set completeopt=menu `  but that didn't seems to change the behavior.

I'm perfectly happy with the workaround using the newer version of vim, initially I thought the issue is more related with jedi-vim directly. However if you would like me to run some more tests in order to pin-point the root cause of course I'll do my best to help. If you prefer this issue to be closed, I'm ok with that too.
